handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslServerContext.java

```diff
@@ -205,4 +218,32 @@ void verify(ReferenceCountedOpenSslEngine engine, X509Certificate[] peerCerts, S
             manager.checkClientTrusted(peerCerts, auth, engine);
         }
     }
+
+    private static final class OpenSslSniHostnameMatcher implements SniHostNameMatcher {
+        private final OpenSslEngineMap engineMap;
+
+        OpenSslSniHostnameMatcher(OpenSslEngineMap engineMap) {
+            this.engineMap = engineMap;
+        }
+
+        @Override
+        public boolean match(long ssl, String hostname) {
+            ReferenceCountedOpenSslEngine engine = engineMap.get(ssl);
+            if (engine != null) {
+                Collection<SNIMatcher> matchers = engine.getSSLParameters().getSNIMatchers();
+                if (matchers != null && !matchers.isEmpty()) {
+                    SNIHostName name = new SNIHostName(hostname);
+                    for (SNIMatcher matcher: matchers) {
+                        // type 0 is for hostname
+                        if (matcher.getType() == 0 && matcher.matches(name)) {
+                            return true;
+                        }
+                    }
+                } else {
+                    return true;
```

Comment:
In this case (engine is null), may be it makes sense at least write warning to the log?